test/unit/org/apache/cassandra/db/guardrails/GuardrailGroupByTest.java

```diff
@@ -0,0 +1,57 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.db.guardrails;
+
+import org.junit.Assert;
+import org.junit.Test;
+
+public class GuardrailGroupByTest extends GuardrailTester
+{
+    private final String query = "SELECT * FROM system_schema.keyspaces GROUP BY keyspace_name";
+
+    private void setGuardrail(boolean enabled)
+    {
+        Guardrails.instance.setGroupByEnabled(enabled);
+    }
+
+    @Test
+    public void checkExplicitlyDisabled() throws Throwable
+    {
+        setGuardrail(false);
+        assertFails(query, "GROUP BY functionality");
+    }
+
+    @Test
+    public void checkEnabled() throws Throwable
+    {
+        setGuardrail(true);
+        try
+        {
+            assertValid(query);
+        }
+        catch (AssertionError ae)
+        {
+            // We expect the query to _warn_ (i.e. fail the assertValid check), but with a non-guardrails warning and
+            // not a failure. Rather than extending the GuardrailsTester framework to handle this case of "warns but not
+            // with a guardrail warning", we just confirm here we saw the acceptable warning we expect but the query was
+            // allowed to run.
+            Assert.assertTrue(ae.getMessage().contains("used without partition key"));
```

Comment:
Good call. I went ahead and switched to using the group by table_name form of the query instead of jumping through these hoops. If we keep seeing this need to differentiate guardrail vs. non-guardrail warnings we can revisit assertValid's functionality.